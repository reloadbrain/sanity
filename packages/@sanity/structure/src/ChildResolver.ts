import {CollectionBuilder, Collection, SerializePath} from './StructureNodes'

export interface ChildResolverOptions {
  index: number
  parentPath?: SerializePath
}

export type ItemChild = CollectionBuilder | Collection | undefined

export interface ChildResolver {
  (itemId: string, parent: Collection, options: ChildResolverOptions):
    | ItemChild
    | Promise<ItemChild>
    | undefined
}
