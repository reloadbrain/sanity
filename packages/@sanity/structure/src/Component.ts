import {SerializeOptions, StructureNode, Serializable} from './StructureNodes'
import {SerializeError, HELP_URL} from './SerializeError'
import {MenuItem, MenuItemBuilder, maybeSerializeMenuItem} from './MenuItem'
import {MenuItemGroup, MenuItemGroupBuilder, maybeSerializeMenuItemGroup} from './MenuItemGroup'

export interface Component extends StructureNode {
  component: Function
  menuItems: MenuItem[]
  menuItemGroups: MenuItemGroup[]
}

export interface ComponentInput extends StructureNode {
  component: Function
  menuItems?: (MenuItem | MenuItemBuilder)[]
  menuItemGroups?: (MenuItemGroup | MenuItemGroupBuilder)[]
}

export interface BuildableList extends Partial<StructureNode> {
  component?: Function
  menuItems?: (MenuItem | MenuItemBuilder)[]
  menuItemGroups?: (MenuItemGroup | MenuItemGroupBuilder)[]
}

export class ComponentBuilder implements Serializable {
  protected spec: BuildableList

  constructor(spec?: ComponentInput) {
    this.spec = spec ? spec : {}
  }

  id(id: string) {
    this.spec.id = id
    return this
  }

  title(title: string) {
    this.spec.title = title
    return this
  }

  component(component: Function) {
    this.spec.component = component
    return this
  }

  menuItems(items: (MenuItem | MenuItemBuilder)[]) {
    this.spec.menuItems = items
    return this
  }

  menuItemGroups(groups: (MenuItemGroup | MenuItemGroupBuilder)[]) {
    this.spec.menuItemGroups = groups
    return this
  }

  serialize(options: SerializeOptions = {path: []}): Component {
    const {id, title, component} = this.spec
    if (typeof id !== 'string' || !id) {
      throw new SerializeError(
        '`id` is required for `component` structure item',
        options.path,
        options.index
      ).withHelpUrl(HELP_URL.ID_REQUIRED)
    }

    if (!component) {
      throw new SerializeError(
        '`component` is required for `component` structure item',
        options.path,
        options.index
      ).withHelpUrl(HELP_URL.ID_REQUIRED)
    }

    return {
      id,
      title,
      type: 'component',
      component,
      menuItems: (this.spec.menuItems || []).map((item, i) =>
        maybeSerializeMenuItem(item, i, options.path)
      ),
      menuItemGroups: (this.spec.menuItemGroups || []).map((item, i) =>
        maybeSerializeMenuItemGroup(item, i, options.path)
      )
    }
  }
}
