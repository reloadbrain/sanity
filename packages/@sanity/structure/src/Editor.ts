import {EditorNode, SerializeOptions, Serializable} from './StructureNodes'
import {SerializeError, HELP_URL} from './SerializeError'
import {SchemaType} from './parts/Schema'

export type PartialEditorNode = {
  id?: string
  title?: string
  options?: {
    id?: string
    type?: string
  }
}

export class EditorBuilder implements Serializable {
  protected spec: PartialEditorNode

  constructor(spec?: EditorNode) {
    this.spec = spec ? spec : {}
  }

  id(id: string): EditorBuilder {
    this.spec.id = id
    return this
  }

  getId() {
    return this.spec.id
  }

  title(title: string): EditorBuilder {
    this.spec.title = title
    return this
  }

  getTitle() {
    return this.spec.title
  }

  documentId(documentId: string): EditorBuilder {
    this.spec.options = {
      ...(this.spec.options || {}),
      id: documentId
    }

    return this
  }

  getDocumentId() {
    return this.spec.options && this.spec.options.id
  }

  schemaType(documentType: SchemaType | string): EditorBuilder {
    this.spec.options = {
      ...(this.spec.options || {}),
      type: typeof documentType === 'string' ? documentType : documentType.name
    }

    return this
  }

  getSchemaType() {
    return this.spec.options && this.spec.options.type
  }

  serialize({path, index, hint}: SerializeOptions = {path: []}): EditorNode {
    const {id, options} = this.spec
    if (typeof id !== 'string' || !id) {
      throw new SerializeError('`id` is required for editor nodes', path, index, hint).withHelpUrl(
        HELP_URL.ID_REQUIRED
      )
    }

    if (!options || !options.id) {
      throw new SerializeError(
        'document id (`id`) is required for editor nodes',
        path,
        id,
        hint
      ).withHelpUrl(HELP_URL.DOCUMENT_ID_REQUIRED)
    }

    if (!options || !options.type) {
      throw new SerializeError(
        'document type (`type`) is required for editor nodes',
        path,
        id,
        hint
      ).withHelpUrl(HELP_URL.DOCUMENT_TYPE_REQUIRED)
    }

    return {
      ...this.spec,
      id,
      type: 'document',
      options: {id: options.id, type: options.type}
    }
  }

  clone(withSpec?: PartialEditorNode) {
    const builder = new EditorBuilder()
    builder.spec = {...this.spec, ...(withSpec || {})}
    return builder
  }
}
