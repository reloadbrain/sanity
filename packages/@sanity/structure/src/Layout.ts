export type Layout = 'default' | 'card' | 'media' | 'detail' | 'block'
export const LayoutOptions = ['default', 'card', 'media', 'detail', 'block']
