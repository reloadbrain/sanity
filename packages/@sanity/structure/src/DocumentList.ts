import {getSerializedChildResolver} from './util/getSerializedChildResolver'
import {client} from './parts/Client'
import {SortItem} from './Sort'
import {EditorBuilder} from './Editor'
import {SerializeError, HELP_URL} from './SerializeError'
import {SerializeOptions, Collection} from './StructureNodes'
import {ChildResolver, ChildResolverOptions, ItemChild} from './ChildResolver'
import {
  GenericListBuilder,
  BuildableGenericList,
  GenericList,
  GenericListInput
} from './GenericList'

const resolveTypeForDocument = (id: string): Promise<string | undefined> => {
  const query = '*[_id in [$documentId, $draftId]]._type'
  const documentId = id.replace(/^drafts\./, '')
  const draftId = `drafts.${documentId}`
  return client.fetch(query, {documentId, draftId}).then(types => types[0])
}

const validateFilter = (spec: PartialDocumentList, options: SerializeOptions) => {
  const filter = spec.options!.filter.trim()

  if (['*', '{'].includes(filter[0])) {
    throw new SerializeError(
      `\`filter\` cannot start with \`${
        filter[0]
      }\` - looks like you are providing a query, not a filter`,
      options.path,
      spec.id,
      spec.title
    ).withHelpUrl(HELP_URL.QUERY_PROVIDED_FOR_FILTER)
  }

  return filter
}

const resolveEditorChildForItem: ChildResolver = (
  itemId: string,
  parent: Collection,
  options: ChildResolverOptions
): ItemChild | Promise<ItemChild> | undefined => {
  const parentItem = parent as DocumentList
  return Promise.resolve(parentItem.schemaTypeName || resolveTypeForDocument(itemId)).then(type =>
    new EditorBuilder()
      .id('editor')
      .documentId(itemId)
      .type(type || '')
  )
}

interface PartialDocumentList extends BuildableGenericList {
  options?: DocumentListOptions
  schemaTypeName?: string
}

export interface DocumentListInput extends GenericListInput {
  options: DocumentListOptions
}

export interface DocumentList extends GenericList {
  options: DocumentListOptions
  resolveChildForItem: ChildResolver
  schemaTypeName?: string
}

interface DocumentListOptions {
  filter: string
  params?: {[key: string]: any}
  defaultOrdering?: SortItem[]
}

export class DocumentListBuilder extends GenericListBuilder<PartialDocumentList> {
  protected spec: PartialDocumentList

  constructor(spec?: DocumentListInput) {
    super()
    this.spec = spec ? spec : {}
  }

  filter(filter: string): DocumentListBuilder {
    this.spec.options = {...(this.spec.options || {}), filter}
    return this
  }

  schemaTypeName(typeName: string): DocumentListBuilder {
    this.spec.schemaTypeName = typeName
    return this
  }

  params(params: {}): DocumentListBuilder {
    this.spec.options = {...(this.spec.options || {filter: ''}), params}
    return this
  }

  defaultOrdering(ordering: SortItem[]): DocumentListBuilder {
    if (!Array.isArray(ordering)) {
      throw new Error('`defaultOrdering` must be an array of order clauses')
    }

    this.spec.options = {...(this.spec.options || {filter: ''}), defaultOrdering: ordering}
    return this
  }

  serialize(options: SerializeOptions = {path: []}): DocumentList {
    if (typeof this.spec.id !== 'string' || !this.spec.id) {
      throw new SerializeError(
        '`id` is required for document lists',
        options.path,
        options.index,
        this.spec.title
      ).withHelpUrl(HELP_URL.ID_REQUIRED)
    }

    if (!this.spec.options || !this.spec.options.filter) {
      throw new SerializeError(
        '`filter` is required for document lists',
        options.path,
        this.spec.id,
        this.spec.title
      ).withHelpUrl(HELP_URL.FILTER_REQUIRED)
    }

    return {
      ...super.serialize(options),
      type: 'documentList',
      schemaTypeName: this.spec.schemaTypeName,
      resolveChildForItem: getSerializedChildResolver(
        this.spec.resolveChildForItem || resolveEditorChildForItem
      ),
      options: {
        ...this.spec.options,
        filter: validateFilter(this.spec, options)
      }
    }
  }
}
