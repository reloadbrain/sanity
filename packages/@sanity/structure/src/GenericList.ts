import {StructureNode, SerializeOptions, Serializable, Child} from './StructureNodes'
import {Layout, layoutOptions} from './Layout'
import {MenuItem, MenuItemBuilder, maybeSerializeMenuItem} from './MenuItem'
import {MenuItemGroup, MenuItemGroupBuilder, maybeSerializeMenuItemGroup} from './MenuItemGroup'
import {IntentChecker} from './Intent'
import {SerializeError} from './SerializeError'

function noChildResolver() {
  return undefined
}

export interface BaseGenericList extends StructureNode {
  defaultLayout?: Layout
  canHandleIntent?: IntentChecker
  child: Child
}

// "POJO"/verbatim-version - end result
export interface GenericList extends BaseGenericList {
  type: string
  menuItems: MenuItem[]
  menuItemGroups: MenuItemGroup[]
}

// Used internally in builder classes to make everything optional
export interface BuildableGenericList extends Partial<BaseGenericList> {
  menuItems?: (MenuItem | MenuItemBuilder)[]
  menuItemGroups?: (MenuItemGroup | MenuItemGroupBuilder)[]
}

// Input version, allows builders and only requires things not inferrable
export interface GenericListInput extends StructureNode {
  id: string
  title: string
  menuItems?: (MenuItem | MenuItemBuilder)[]
  menuItemGroups?: (MenuItemGroup | MenuItemGroupBuilder)[]
  defaultLayout?: Layout
  canHandleIntent?: IntentChecker
  child?: Child
}

export abstract class GenericListBuilder<L extends BuildableGenericList> implements Serializable {
  protected intentChecker?: IntentChecker
  protected spec: L = {} as L

  id(id: string) {
    this.spec.id = id
    return this
  }

  title(title: string) {
    this.spec.title = title
    return this
  }

  defaultLayout(layout: Layout) {
    this.spec.defaultLayout = layout
    return this
  }

  menuItems(items: (MenuItem | MenuItemBuilder)[]) {
    this.spec.menuItems = items
    return this
  }

  menuItemGroups(groups: (MenuItemGroup | MenuItemGroupBuilder)[]) {
    this.spec.menuItemGroups = groups
    return this
  }

  child(resolver: Child) {
    this.spec.child = resolver
    return this
  }

  canHandleIntent(checker: IntentChecker) {
    this.intentChecker = checker
    return this
  }

  serialize(options: SerializeOptions = {path: []}): GenericList {
    const id = this.spec.id || ''
    const path = options.path

    const defaultLayout = this.spec.defaultLayout
    if (defaultLayout && !layoutOptions.includes(defaultLayout)) {
      throw new SerializeError(
        `\`layout\` must be one of ${layoutOptions.map(item => `"${item}"`).join(', ')}`,
        path,
        id || options.index,
        this.spec.title
      )
    }

    return {
      id,
      title: this.spec.title,
      type: 'genericList',
      defaultLayout,
      child: this.spec.child || noChildResolver,
      canHandleIntent: this.intentChecker,
      menuItems: (this.spec.menuItems || []).map((item, i) =>
        maybeSerializeMenuItem(item, i, path)
      ),
      menuItemGroups: (this.spec.menuItemGroups || []).map((item, i) =>
        maybeSerializeMenuItemGroup(item, i, path)
      )
    }
  }
}
