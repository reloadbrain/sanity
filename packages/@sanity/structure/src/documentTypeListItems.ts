import memoizeOne from 'memoize-one'
import {Schema, defaultSchema, SchemaType} from './parts/Schema'
import {dataAspects, DataAspectsResolver} from './parts/DataAspects'
import {ListItemBuilder} from './ListItem'
import {DocumentListBuilder} from './DocumentList'
import {MenuItemBuilder} from './MenuItem'
import {DEFAULT_ORDERING_OPTIONS, DEFAULT_SELECTED_ORDERING_OPTION} from './Sort'
import {getPlusIcon, getSortIcon, getListIcon} from './parts/Icon'
import {EditorBuilder} from './Editor'

const PlusIcon = getPlusIcon()
const SortIcon = getSortIcon()
const ListIcon = getListIcon()

const getDataAspectsForSchema: (schema: Schema) => DataAspectsResolver = memoizeOne(dataAspects)

export function getDocumentTypeListItems(schema: Schema = defaultSchema): ListItemBuilder[] {
  const resolver = getDataAspectsForSchema(schema)
  const types = resolver.getInferredTypes()
  return types.map(name => getDocumentTypeListItem(name, schema))
}

export function getDocumentTypeListItem(
  name: string,
  schema: Schema = defaultSchema
): ListItemBuilder {
  const type = schema.get(name)
  const resolver = getDataAspectsForSchema(schema)
  const title = resolver.getDisplayName(name)
  return new ListItemBuilder()
    .id(name)
    .title(title)
    .schemaType(type)
    .child(getDocumentList(name, title, type))
}

function getDocumentList(name: string, title: string, type: SchemaType): DocumentListBuilder {
  return new DocumentListBuilder()
    .id(name)
    .title(title)
    .filter('_type == $type')
    .params({type: name})
    .schemaTypeName(name)
    .defaultOrdering(DEFAULT_SELECTED_ORDERING_OPTION.by)
    .canHandleIntent(
      (intentName, params): boolean =>
        Boolean(intentName === 'edit' && params && params.id) ||
        Boolean(intentName === 'create' && params && params.type === name)
    )
    .menuItemGroups([
      {id: 'sorting', title: 'Sort'},
      {id: 'layout', title: 'Layout'},
      {id: 'actions', title: 'Actions'}
    ])
    .childResolver((documentId: string) =>
      new EditorBuilder()
        .id('editor')
        .type(name)
        .documentId(documentId)
    )
    .menuItems([
      // Create new (from action button)
      new MenuItemBuilder()
        .title(`Create new ${title}`)
        .icon(PlusIcon)
        .intent({type: 'create', params: {type: name}})
        .showAsAction({whenCollapsed: true}),

      // Sort by <Y>
      ...getOrderingMenuItemsForSchemaType(type),

      // Display as <Z>
      new MenuItemBuilder()
        .group('layout')
        .title('List')
        .icon(ListIcon)
        .action('setLayout')
        .params({layout: 'default'}),

      new MenuItemBuilder()
        .group('layout')
        .title('Details')
        .icon(ListIcon)
        .action('setLayout')
        .params({layout: 'detail'}),

      // Create new (from menu)
      new MenuItemBuilder()
        .group('actions')
        .title('Create newâ€¦')
        .icon(PlusIcon)
        .intent({type: 'create', params: {type: name}})
    ])
}

function getOrderingMenuItemsForSchemaType(type: SchemaType) {
  return (type.orderings
    ? type.orderings.concat(DEFAULT_ORDERING_OPTIONS)
    : DEFAULT_ORDERING_OPTIONS
  ).map(ordering =>
    new MenuItemBuilder()
      .group('sorting')
      .title(`Sort by ${ordering.title}`)
      .icon(SortIcon)
      .action('setSortOrder')
      .params({by: ordering.by})
  )
}
