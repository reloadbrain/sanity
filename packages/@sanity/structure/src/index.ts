import {ListBuilder, ListInput} from './List'
import {getDocumentTypeListItems, getDocumentTypeListItem} from './documentTypeListItems'
import {MenuItemBuilder, MenuItem} from './MenuItem'
import {ListItemBuilder, ListItem} from './ListItem'
import {MenuItemGroup, MenuItemGroupBuilder} from './MenuItemGroup'
import {DocumentListBuilder, DocumentListInput} from './DocumentList'
import {EditorBuilder} from './Editor'
import {EditorNode} from './StructureNodes'
import {SerializeError} from './SerializeError'
import {ComponentInput, ComponentBuilder} from './Component'
import {DocumentListItemBuilder, DocumentListItemInput} from './DocumentListItem'

const StructureBuilder = {
  documentTypeListItem: getDocumentTypeListItem,
  documentTypeListItems: getDocumentTypeListItems,
  list: (spec?: ListInput) => new ListBuilder(spec),
  listItem: (spec?: ListItem) => new ListItemBuilder(spec),
  editor: (spec?: EditorNode) => new EditorBuilder(spec),
  menuItem: (spec?: MenuItem) => new MenuItemBuilder(spec),
  menuItemGroup: (spec?: MenuItemGroup) => new MenuItemGroupBuilder(spec),
  documentList: (spec?: DocumentListInput) => new DocumentListBuilder(spec),
  documentListItem: (spec?: DocumentListItemInput) => new DocumentListItemBuilder(spec),

  component: (spec?: ComponentInput | Function) => {
    return typeof spec === 'function'
      ? new ComponentBuilder().component(spec)
      : new ComponentBuilder(spec)
  }
}

export {StructureBuilder, SerializeError}
