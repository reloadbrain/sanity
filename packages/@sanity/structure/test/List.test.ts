import {StructureBuilder as S} from '../src'
import serializeStructure from './util/serializeStructure'

const noop = () => {
  /* intentional noop */
}

test('builds lists with only required props', () => {
  expect(S.list({id: 'foo', title: 'Foo'}).serialize()).toMatchSnapshot()
})

test('throws if no id is set', () => {
  expect(() => S.list().serialize()).toThrowErrorMatchingSnapshot()
})

test('builds lists with ID and title through setters', () => {
  expect(
    S.list({id: 'books', title: 'Books'})
      .id('authors')
      .title('Authors')
      .serialize()
  ).toMatchSnapshot()
})

test('builds lists with specific items', () => {
  expect(
    S.list()
      .id('books')
      .items([{id: 'asoiaf-wow', title: 'The Winds of Winter'}])
      .serialize()
  ).toMatchSnapshot()
})

test('builds lists where items are specified using builder', () => {
  expect(
    S.list()
      .id('books')
      .items([S.listItem({id: 'foo', title: 'Foo'})])
      .serialize()
  ).toMatchSnapshot()
})

test('builds lists with layout', () => {
  expect(
    S.list()
      .id('books')
      .defaultLayout('card')
      .serialize()
  ).toMatchSnapshot()
})

test('default child resolver can resolve directly to node', done => {
  const item = {
    id: 'asoiaf-wow',
    title: 'The Winds of Winter',
    child: {id: 'editor', type: 'document', options: {id: 'wow', type: 'book'}}
  }

  const list = S.list()
    .id('books')
    .items([item])
    .serialize()

  const context = {parent: list, index: 0}
  serializeStructure(list.child, context, [item.id, context]).subscribe(child => {
    expect(child).toEqual(item.child)
    done()
  })
})

test('default child resolver can resolve through promise', done => {
  const child = {id: 'editor', type: 'document', options: {id: 'wow', type: 'book'}}
  const item = {
    id: 'asoiaf-wow',
    title: 'The Winds of Winter',
    child: () => Promise.resolve(child)
  }

  const list = S.list()
    .id('books')
    .items([item])
    .serialize()

  const context = {parent: list, index: 0}
  serializeStructure(list.child, context, [item.id, context]).subscribe(itemChild => {
    expect(itemChild).toEqual(child)
    done()
  })
})

test('can provide custom child resolver', done => {
  const list = S.list()
    .id('books')
    .items([{id: 'today', title: 'Today'}])
    .child(() => ({
      id: 'editor',
      type: 'editor',
      options: {id: new Date().toISOString().slice(0, 10), type: 'gallery'}
    }))
    .serialize()

  const context = {parent: list, index: 0}
  serializeStructure(list.child, context, ['today', context]).subscribe(child => {
    expect(child).toHaveProperty('options.id', new Date().toISOString().slice(0, 10))
    done()
  })
})

test('can resolve undefined child', done => {
  const list = S.list()
    .id('books')
    .items([{id: 'today', title: 'Today'}])
    .child(() => undefined)
    .serialize()

  const context = {parent: list, index: 0}
  serializeStructure(list.child, context, ['today', context]).subscribe(child => {
    expect(child).toBeUndefined()
    done()
  })
})

test('can set menu items', () => {
  expect(
    S.list()
      .id('yeah')
      .menuItems([{title: 'Print', action: noop}])
      .serialize()
  ).toMatchSnapshot()
})

test('can set menu items with builder', () => {
  expect(
    S.list()
      .id('yeah')
      .menuItems([
        S.menuItem()
          .title('Purge')
          .action(noop)
      ])
      .serialize()
  ).toMatchSnapshot()
})

test('can set menu item groups', () => {
  expect(
    S.list()
      .id('yeah')
      .menuItems([{title: 'Print', action: noop, group: 'old-school'}])
      .menuItemGroups([{title: 'Old-school', id: 'old-school'}])
      .serialize()
  ).toMatchSnapshot()
})

test('can set menu items groups with builder', () => {
  expect(
    S.list()
      .id('yeah')
      .menuItems([
        S.menuItem()
          .title('Print')
          .action(noop)
          .group('old-school')
      ])
      .menuItemGroups([
        S.menuItemGroup()
          .id('old-school')
          .title('Old-school')
      ])
      .serialize()
  ).toMatchSnapshot()
})

test('can set intent handler check', () => {
  const handler = () => false
  expect(
    S.list()
      .id('yeah')
      .canHandleIntent(handler)
      .serialize()
  ).toMatchObject({canHandleIntent: handler})
})
