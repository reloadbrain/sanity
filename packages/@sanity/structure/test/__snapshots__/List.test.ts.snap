// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builds lists where items are specified using builder 1`] = `
Object {
  "canHandleIntent": undefined,
  "defaultLayout": undefined,
  "id": "books",
  "items": Array [
    Object {
      "child": undefined,
      "id": "foo",
      "schemaType": undefined,
      "title": "Foo",
    },
  ],
  "menuItemGroups": Array [],
  "menuItems": Array [],
  "resolveChildForItem": [Function],
  "title": undefined,
  "type": "list",
}
`;

exports[`builds lists with ID and title through setters 1`] = `
Object {
  "canHandleIntent": undefined,
  "defaultLayout": undefined,
  "id": "authors",
  "items": Array [],
  "menuItemGroups": Array [],
  "menuItems": Array [],
  "resolveChildForItem": [Function],
  "title": "Authors",
  "type": "list",
}
`;

exports[`builds lists with layout 1`] = `
Object {
  "canHandleIntent": undefined,
  "defaultLayout": "card",
  "id": "books",
  "items": Array [],
  "menuItemGroups": Array [],
  "menuItems": Array [],
  "resolveChildForItem": [Function],
  "title": undefined,
  "type": "list",
}
`;

exports[`builds lists with only required props 1`] = `
Object {
  "canHandleIntent": undefined,
  "defaultLayout": undefined,
  "id": "foo",
  "items": Array [],
  "menuItemGroups": Array [],
  "menuItems": Array [],
  "resolveChildForItem": [Function],
  "title": "Foo",
  "type": "list",
}
`;

exports[`builds lists with specific items 1`] = `
Object {
  "canHandleIntent": undefined,
  "defaultLayout": undefined,
  "id": "books",
  "items": Array [
    Object {
      "id": "asoiaf-wow",
      "title": "The Winds of Winter",
    },
  ],
  "menuItemGroups": Array [],
  "menuItems": Array [],
  "resolveChildForItem": [Function],
  "title": undefined,
  "type": "list",
}
`;

exports[`can set menu item groups 1`] = `
Object {
  "canHandleIntent": undefined,
  "defaultLayout": undefined,
  "id": "yeah",
  "items": Array [],
  "menuItemGroups": Array [
    Object {
      "id": "old-school",
      "title": "Old-school",
    },
  ],
  "menuItems": Array [
    Object {
      "action": [Function],
      "group": "old-school",
      "title": "Print",
    },
  ],
  "resolveChildForItem": [Function],
  "title": undefined,
  "type": "list",
}
`;

exports[`can set menu items 1`] = `
Object {
  "canHandleIntent": undefined,
  "defaultLayout": undefined,
  "id": "yeah",
  "items": Array [],
  "menuItemGroups": Array [],
  "menuItems": Array [
    Object {
      "action": [Function],
      "title": "Print",
    },
  ],
  "resolveChildForItem": [Function],
  "title": undefined,
  "type": "list",
}
`;

exports[`can set menu items groups with builder 1`] = `
Object {
  "canHandleIntent": undefined,
  "defaultLayout": undefined,
  "id": "yeah",
  "items": Array [],
  "menuItemGroups": Array [
    Object {
      "id": "old-school",
      "title": "Old-school",
    },
  ],
  "menuItems": Array [
    Object {
      "action": [Function],
      "group": "old-school",
      "title": "Print",
    },
  ],
  "resolveChildForItem": [Function],
  "title": undefined,
  "type": "list",
}
`;

exports[`can set menu items with builder 1`] = `
Object {
  "canHandleIntent": undefined,
  "defaultLayout": undefined,
  "id": "yeah",
  "items": Array [],
  "menuItemGroups": Array [],
  "menuItems": Array [
    Object {
      "action": [Function],
      "title": "Purge",
    },
  ],
  "resolveChildForItem": [Function],
  "title": undefined,
  "type": "list",
}
`;

exports[`throws if no id is set 1`] = `"\`id\` is required for lists"`;
